### PR TITLE
Java ITs: Remove "bulk_delete" orchestrator call

### PR DESCRIPTION
**Pull request by gregory-paidis-sonarsource**
Part of #7557 

- CAYC: Remove some useless imports
- CAYC: Changed some `Before` to `BeforeClass` annotations (running once per _class_ instead of per _file_)
- Remove `bulk_delete` call from `orchestrator.reset()` calls
- vbnet+shared work out-of-the-box
- C# needed `projectKey` de-collisioning for `ExternalIssuesTest` and `IncrementalAnalysisTest`
- Added an overload method for `csharp.Tests.analyzeProject` to support different `projectName` and `projectDir`

### CODE REVIEW

File: its/src/test/java/com/sonar/it/csharp/AnalysisWarningsTest.java [hunk 1]
```diff
@@ -37,8 +37,8 @@ public class AnalysisWarningsTest {
   @Rule
   public TemporaryFolder temp = TestUtils.createTempFolder();
 
-  @Before
-  public void init() {
+  @BeforeClass
+  public static void init() {
```
Comment (gregory-paidis-sonarsource) [inline its/src/test/java/com/sonar/it/csharp/AnalysisWarningsTest.java:41]: These changes are pretty irrelevant, but there is no reason to run the `server is not null` check per test, since it should be started once in total anyways, so `BeforeClass` is semantically more correct IMO.

File: its/src/test/java/com/sonar/it/csharp/Tests.java [hunk 2]
```diff
@@ -97,18 +99,20 @@ public static void deleteLocalCache() {
     TestUtils.deleteLocalCache();
   }
 
-  static BuildResult analyzeProject(TemporaryFolder temp, String projectName, @Nullable String profileKey, String... keyValues) throws IOException {
-    Path projectDir = Tests.projectDir(temp, projectName);
+  static BuildResult analyzeProject(String projectDir, TemporaryFolder temp, String projectName, @Nullable String profileKey, String... keyValues) throws IOException {
```
[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply its/src/test/java/com/sonar/it/csharp/Tests.java:102]: After this, `ExternalIssuesTest` and `IncrementalAnalysisTest` look busy, but it's mostly boring refactoring of the variables to pass them correctly to orchestrator.

File: its/src/test/java/com/sonar/it/csharp/Tests.java [hunk 3]
```diff
@@ -97,18 +99,20 @@ public static void deleteLocalCache() {
...
     ORCHESTRATOR.executeBuild(beginStep);
+    TestUtils.runMSBuild(ORCHESTRATOR, projectFullPath, "/t:Restore,Rebuild");
+    return ORCHESTRATOR.executeBuild(TestUtils.createEndStep(projectFullPath));
+  }
 
-    TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Restore,Rebuild");
-
-    return ORCHESTRATOR.executeBuild(TestUtils.createEndStep(projectDir));
+  static BuildResult analyzeProject(TemporaryFolder temp, String projectName, @Nullable String profileKey, String... keyValues) throws IOException {
+    return analyzeProject(projectName, temp, projectName, profileKey, keyValues);
```
Comment (gregory-paidis-sonarsource) [inline its/src/test/java/com/sonar/it/csharp/Tests.java:115]: This is the old overload, it calls the new one with the same `projectName` at two argument positions.

File: its/src/test/java/com/sonar/it/shared/TestUtils.java [hunk 4]
```diff
@@ -242,18 +242,6 @@ public static void reset(Orchestrator orchestrator) {
...
-    SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
-    String currentDateTime = formatter.format(new Date());
-
-    LOG.info("TEST SETUP: deleting projects analyzed before: " + currentDateTime);
-    orchestrator.getServer()
-      .newHttpCall("/api/projects/bulk_delete")
-      .setAdminCredentials()
-      .setMethod(HttpMethod.POST)
-      .setParams("analyzedBefore", currentDateTime)
-      .execute();
```
Comment (gregory-paidis-sonarsource) [inline its/src/test/java/com/sonar/it/shared/TestUtils.java:256]: 📡  ---> 🛰️  ---> 🚀

File: its/src/test/java/com/sonar/it/csharp/IncrementalAnalysisTest.java [hunk 5]
```diff
@@ -22,6 +22,7 @@
 import com.sonar.it.shared.TestUtils;
 import com.sonar.orchestrator.build.BuildResult;
 import com.sonar.orchestrator.build.ScannerForMSBuild;
+
```
Comment (pavel-mikula-sonarsource) [inline its/src/test/java/com/sonar/it/csharp/IncrementalAnalysisTest.java:25]: There's something wrong with your IntelliJ setup. When I apply formatting, these empty lines get removed.

See https://github.com/SonarSource/sonar-developer-toolset#code-style-configuration-for-intellij

File: its/src/test/java/com/sonar/it/vbnet/UnitTestResultsTest.java [hunk 6]
```diff
@@ -20,14 +20,9 @@
...
-import com.sonar.orchestrator.Orchestrator;
-import com.sonar.orchestrator.build.ScannerForMSBuild;
 import java.io.IOException;
-import java.nio.file.Path;
-import org.junit.Before;
-import org.junit.ClassRule;
-import org.junit.Rule;
-import org.junit.Test;
+
+import org.junit.*;
```
Comment (pavel-mikula-sonarsource) [inline its/src/test/java/com/sonar/it/vbnet/UnitTestResultsTest.java:25]: @andrei-epure-sonarsource Do we generally use this trick? I haven't seen it elsewhere AFAICT.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Round 1, I like your changes too much... I need more :D

File: its/src/test/java/com/sonar/it/csharp/AutoGeneratedTest.java [hunk 7]
```diff
@@ -37,8 +37,8 @@ public class AutoGeneratedTest {
   @ClassRule
   public static TemporaryFolder temp = TestUtils.createTempFolder();
 
-  @Before
-  public void init() {
+  @BeforeClass
```
[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply its/src/test/java/com/sonar/it/csharp/AutoGeneratedTest.java:40]: List of modified files went from 5 to 40 :)
but we are in the red in the lines of code (minus 10), and most of the changes are trivial renames+move-arounds.

File: its/src/test/java/com/sonar/it/vbnet/Tests.java [hunk 8]
```diff
@@ -77,6 +77,10 @@ public static void deleteLocalCache() {
     TestUtils.deleteLocalCache();
   }
 
+  static BuildResult analyzeProject(TemporaryFolder temp, String projectName) throws IOException {
```
Comment (pavel-mikula-sonarsource) [inline its/src/test/java/com/sonar/it/vbnet/Tests.java:80]: To keep it aligned with CSharp (we don't need all those overloads, but the existing ones should match)
```suggestion
  static BuildResult analyzeProject(TemporaryFolder temp, String projectDir) throws IOException {
```

File: its/src/test/java/com/sonar/it/csharp/MultiTargetAppTest.java [hunk 9]
```diff
@@ -32,20 +33,21 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
 public class MultiTargetAppTest {
-  @Rule
-  public TemporaryFolder temp = TestUtils.createTempFolder();
 
-  @Before
-  public void init() {
-    TestUtils.reset(ORCHESTRATOR);
+  private static BuildResult buildResult;
```
Comment (pavel-mikula-sonarsource) [inline its/src/test/java/com/sonar/it/csharp/MultiTargetAppTest.java:37]: Why this change with buildResult?

File: its/src/test/java/com/sonar/it/csharp/ProjectLevelDuplicationTest.java [hunk 10]
```diff
@@ -36,18 +37,17 @@
...
+  @ClassRule
+  public static TemporaryFolder temp = TestUtils.createTempFolder();
 
-  @Before
-  public void init() {
-    TestUtils.reset(ORCHESTRATOR);
+  @BeforeClass
+  public static void init() throws IOException {
+    TestUtils.initLocal(ORCHESTRATOR);
+    Tests.analyzeProject(temp, "ProjectLevelDuplication");
```
Comment (pavel-mikula-sonarsource) [inline its/src/test/java/com/sonar/it/csharp/ProjectLevelDuplicationTest.java:46]: I'm not sure why this would happen in @BeforeClass and not the UT. There's just one UT anyway

File: its/src/test/java/com/sonar/it/csharp/RuleParameterCustomizationTest.java [hunk 11]
```diff
@@ -37,28 +38,24 @@ public class RuleParameterCustomizationTest {
...
-  public void init() {
-    TestUtils.reset(ORCHESTRATOR);
+  @ClassRule
+  public static TemporaryFolder temp = TestUtils.createTempFolder();
 
+  @BeforeClass
+  public static void init() throws IOException {
+    TestUtils.initLocal(ORCHESTRATOR);
     provisionProject();
+    Tests.analyzeProject(temp, CustomParametersProjectName);
```
Comment (pavel-mikula-sonarsource) [inline its/src/test/java/com/sonar/it/csharp/RuleParameterCustomizationTest.java:48]: Same question here

File: its/src/test/java/com/sonar/it/shared/Tests.java [hunk 12]
```diff
@@ -63,6 +63,10 @@ public static Path projectDir(TemporaryFolder temp, String projectName) throws I
     return tmpProjectDir;
   }
 
+  static BuildResult analyzeProject(TemporaryFolder temp, String projectName) throws IOException {
```
Comment (pavel-mikula-sonarsource) [inline its/src/test/java/com/sonar/it/shared/Tests.java:66]: Parameter names should get aligned with CSharp version here too

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Almost done :tm:

File: its/src/test/java/com/sonar/it/csharp/MultiTargetAppTest.java (hunk 9 above)
Comment (gregory-paidis-sonarsource) [reply its/src/test/java/com/sonar/it/csharp/MultiTargetAppTest.java]: This is the same with the `BeforeClass` ones. I thought we wanted every `analyzeProject` to be in BeforeClass.
I reverted the changes on these classes.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM
